proposals/reference-types/Overview.md

```diff
@@ -0,0 +1,204 @@
+# Reference Types for WebAssembly
+
+TODO: more text, motivation, explanation
+
+## Introduction
+
+Motivation:
+
+* Easier and more efficient interop with host environment
+  - allow host references to be represented directly by type `anyref`
+  - without having to go through tables, allocating slots, and maintaining index bijections at the boundaries
+
+* Basic manipulation of tables inside Wasm
+  - allow representing data structures containing references
+by repurposing tables as a general memory for opaque data types
+  - allow manipulating function tables from within Wasm.
+
+* Set the stage for later additions:
+
+  - Typed function references (see below)
+  - Exception references (see exception handling proposal)
+  - A smoother transition path to GC (see GC proposal)
+
+Get the most important parts soon!
+
+Summary:
+
+* Add a new type `anyref` that can be used as both a value type and a table element type.
+
+* Also allow `anyfunc` as a value type.
+
+* Introduce instructions to get and set table slots.
+
+* Allow multiple tables.
+
+Notes:
+
+* This extension does not imply GC by itself, only if host refs are GCed pointers!
+
+* Reference types are *opaque*, i.e., their value is abstract and they cannot be stored into linear memory. Tables are used as the equivalent.
+
+
+## Language Extensions
+
+Typing extensions:
+
+* Introduce `anyref`, `anyfunc`, and `nullref` as a new class of *reference types*.
+  - `reftype ::= anyref | anyfunc | nullref`
```

Comment:
I wonder if the `nullref` type is really necessary.  IIUC, having this as a normal value type means we have to handle this degenerate type in locals, globals (and so `WebAssembly.Global`s) and function signatures.  Adding separate "expression" vs. "value" type sets just to avoid these degenerate case would also be additional complexity.  Instead of having `ref.null` producing `nullref`, I was imagining a `null $t` which produced the desired reference type `$t` without any subtyping.

Reply:
It just seemed natural. Can you elaborate in what sense it would introduce a "degenerate" case? AFAICS it falls out for free from the way we will have to uniformly allow for subtyping anyway in all these cases.

Reply:
If locals, globals, parameters, `WebAssembly.Global` etc have type `nullref`; this will actually require a bit of implementation work (extending the call ABI (trivially), deciding how to represent `nullref` locals (and whether to optimize them to rematerialize null when `get_local`d or whether to actually store a null value on the stack)) and noone in practice will use `nullref` for any of these purposes.  It seems like a validation detail.

Reply:
Hope you don't mind my questions, but why wouldn't you simply implement them like all reference types? I agree that these cases would be mostly useless in practice, so no need to bother with optimisations or other special casing. AFAICS, the only thing that's strictly necessary is adding one more case to the implementation of ToWebAssemblyValue, which will get several new cases anyway.

For background, another technical reason for having nullref (at least later) is that some things are easier if the subtyping relation forms a lattice, especially when you have structural subtyping on function types (because of contra-variance). There are other places where nullref might come in handy, e.g. as a way to represent nullable types via unions (see comment below).

Anyway, I'm not tied to providing it explicitly, we could just make it an "unnamed" type, i.e., it's part of the validation semantics but you can't use it in declarations. But it feels somewhat incomplete that way.

Reply:
Yeah, we'd just lower it to a reference.  But consider the [js-types](https://github.com/WebAssembly/js-types) proposal; it'd have to specifically mention "null" and that would look kinda warty.  Given the nullable vs. non-nullable types we're talking about separately, considering null to be an operation only available to the nullable ones seems natural.

Also, would `nulltype` be a suitable bottom of the lattice given the non-nullable reference types?

Reply:
Okay, degraded it to an "internal" type. It's true that the JS API would also want to expose `nullref` if we made it available, but I'm not sure why that would be a wart. I don't understand your second point; it follows naturally from the subtype relation that `ref.null` is only available for nullable types.

You are right, `nullref` wouldn't be the bottom type.

Reply:
Oops, my second point was w.r.t the alternative proposal to not have a separate `nullref` type and instead have `t.null` be an operation for nullable `t`.  But with the other changes, this is a pretty minute detail at this point, so happy to discuss separately and after more experience.